feat: add mediator middleware type for play()

ldayananda opened the pull request:
## Description
This will allow middleware to interact with calls to play() from the tech. This will require a method of indicating to middleware previously run that a middleware down the chain has terminated or stopped execution. 

## Specific Changes proposed
- Adds middleware `mediator` method that runs middleware from the player to the tech and a second time back up to the player. This category was created because `play` is both a setter(changes the playback state) and a getter(gets a native play promise if available). This also has the ability to tell whether a middleware has terminated before reaching the tech.
- Adds a `middleware.TERMINATOR` sentinel value that is available on the `videojs` object
- Adds `play` to the allowedMediators
- Adds `paused` to the allowedGetters
- Adds a sandbox example of a play mediator middleware

## Checklist
- [x] Feature implemented / Bug fixed
- [x] unit tests added
- [x] documentation added
- [x] guide added/updated
- [x] verified in browser
- [x] example created
- [x] Reviewed by Two Core Contributors

ldayananda commented on src/js/player.js:
```diff
@@ -1632,6 +1632,9 @@ class Player extends Component {
       if (method in middleware.allowedSetters) {
         return middleware.set(this.middleware_, this.tech_, method, arg);
+
+      } else if (method in middleware.allowedMediators) {
```
`player.play()` seems to be done via `techGet` for the html5 tech, but it's conceivable that a future mediator will use `techCall`

gkatsev replied:
maybe we should just make a `techCallAndGet` for stuff like play?

ldayananda replied:
That might make sense; I don't really strongly about it one way or another.

gkatsev replied:
Let's get everything else ready and then come back to this.

gkatsev replied:
eh, I think we can just leave it like this.

ldayananda replied:
Yeah, doesn't feel like there's a pressing need to change this yet

gkatsev commented on src/js/tech/middleware.js:
```diff
@@ -2,6 +2,8 @@ import { assign } from '../utils/obj.js';
 
 const middlewares = {};
 
+export const TERMINATOR = 'TERMINATOR';
```
we should make this be an object instance rather than a string since objects only equal themselves. Then the string will also be a valid value.

ldayananda replied:
what do you mean by "the string will also be a valid value"

ldayananda replied:
Did you mean change this to `export const TERMINATOR = () => { return 'TERMINATOR'; }` and then here:

https://github.com/ldayananda/video.js/blob/52011c3bc175897e735ec5018159de222db0980c/src/js/video.js#L752

change it to be `videojs.TERMINATOR = TERMINATOR()` ?

gkatsev replied:
nope, literally just `TERMINATOR = {}`

ldayananda replied:
Oh, you meant "the string will also be a valid value" to mean that an allowed method could return the string 'TERMINATOR'! 👍  will do

misteroneill commented on src/js/video.js:
```diff
@@ -749,5 +749,7 @@ videojs.dom = Dom;
  */
 videojs.url = Url;
 
+videojs.TERMINATOR = TERMINATOR;
```
If we have access to `Object.defineProperty` we should probably use it here to prevent overwriting this value. Others may think that's overkill, though.

ldayananda replied:
Seems reasonable, I'll do that

ldayananda commented on src/js/video.js:
```diff
@@ -749,7 +749,9 @@ videojs.dom = Dom;
 videojs.url = Url;
 
-videojs.TERMINATOR = TERMINATOR;
+Object.defineProperty(videojs, 'TERMINATOR', {
```
@misteroneill does this seem right?

gkatsev commented on src/js/video.js:
```diff
@@ -749,5 +749,9 @@ videojs.dom = Dom;
  */
 videojs.url = Url;
 
+Object.defineProperty(videojs, 'TERMINATOR', {
```
this will break in IE8 :)

Also, I think it's reasonable to make it enumerable and it might be nice to make writeable false be explicit.

ldayananda replied:
Is there an example somewhere else in the project of how to do this for IE8 as well? I tried to find one earlier and came up empty.

🆗  explicit `writeable: false` and `enumerable: true` sound good to me

gkatsev replied:
something like
```js
if (Object.defineProperty && !browser.IS_IE8) {
  Object.defineProperty(...);
} else {
  videojs.TERMINATOR = TERMINATOR;
}
```

gkatsev commented on src/js/video.js:
```diff
@@ -749,5 +749,9 @@ videojs.dom = Dom;
  */
 videojs.url = Url;
 
+Object.defineProperty(videojs, 'TERMINATOR', {
```
should this be directly on `videojs`? Maybe namespaced `videojs.middleware.TERMINATOR`?

ldayananda replied:
I like the idea of it being namespaced to `videojs.middleware.TERMINATOR`

gkatsev commented on src/js/tech/middleware.js:
```diff
@@ -31,20 +34,41 @@ export function set(middleware, tech, method, arg) {
 };
 
+export const allowedMediators = {
+  play: 1
```
are we adding `pause` here as well?

ldayananda replied:
Oh, I should. Thanks for catching that

gkatsev commented on src/js/tech/middleware.js:
```diff
@@ -53,6 +77,14 @@ function middlewareIterator(method) {
+  for (let i = mws.length - 1; i >= 0; i--) {
+    const mw = mws[i];
+
+    mw[method](value, terminated);
```
can we swap the argument order so it's similar to a node-style callback?

ldayananda replied:
Good idea, we can do that.

gkatsev commented on src/js/big-play-button.js:
```diff
@@ -59,7 +59,7 @@ class BigPlayButton extends Button {
     const playFocus = () => playToggle.focus();
 
-    if (isPromise(playPromise)) {
+    if (playPromise && isPromise(playPromise)) {
```
`isPromise` should be doing a null check as well, no, this shouldn't be strictly necessary.

ldayananda replied:
I did notice that `isPromise` was checking for `undefined` specifically. I thought that was for a specific reason, but maybe I was wrong?

gkatsev replied:
if `play()` is called and no promise was returned, we end up with `undefined`. Though, we'll never go into this if statement if `playPromise` doesn't duck-type into a promise.
I'm ok with leaving it if you want, but it shouldn't be necessary.

ldayananda replied:
Ah, while I was testing, if the play was terminated then I saw an error coming from this `if` statement. Perhaps the right way is to return something other than `null` in the cases where a middleware terminates.

gkatsev replied:
Ah, I see. Maybe we should just loosen the check in `isPromise`

gkatsev commented on src/js/video.js:
```diff
@@ -751,8 +751,8 @@ videojs.url = Url;
 if (!browser.IS_IE8 && Object.defineProperty) {
-  Object.defineProperty(videojs, 'middleware.TERMINATOR', {
-    value: TERMINATOR,
+  Object.defineProperty(videojs, 'middleware', {
```
you'd want defineProperty each step because now you could overwrite `videojs.middleware.TERMINATOR = false`

ldayananda replied:
Good point

ldayananda commented on docs/guides/middleware.md:
```diff
@@ -0,0 +1,116 @@
+# Middleware
```
This will have to be rebased once the current middleware guide is merged

gkatsev commented:
Looking at this version of the middleware guide (https://deploy-preview-4868--videojs-docs.netlify.com/tutorial-middleware.html), I realized we forgot to document the "special" `setTech` method. Also, looks like we say we have to return an object with these methods but an instance of a class or prototype or w/e will work too. But we can do this as part of a separate PR, probably.
Also, the package-lock has conflicts, yay! (sorry)

gkatsev commented on sandbox/middleware-play.example:
```diff
@@ -0,0 +1,184 @@
+<!DOCTYPE html>
```
this file should be `middleware-play.html.example`

gkatsev commented on sandbox/middleware-play.example:
```diff
@@ -0,0 +1,184 @@
+      margin: 2em 1em;
+    }
+
+    .vjs-progress-control.vjs-control .vjs-play-progress.vjs-slider-bar.terminated {
```
why not look for the `terminated` class on the player instead?

gkatsev replied:
```css
.terminated .vjs-process-control .vjs-play-progress
```
Then below (line 44), you can just do `player.addClass('terminated')` instead of going through the entire child heirarchy to get the proper element. Or something like that.

ldayananda commented:
@gkatsev I think I've now addressed all your comments. I agree that the corrections to the middleware guide can be done in a separate PR. And I removed the package-lock so we should be good now 😄

gkatsev commented on src/js/tech/middleware.js:
```diff
@@ -26,15 +26,28 @@ export function setTech(middleware, tech) {
+/**
+ * Calls a getter on the tech first, through each middleware
+ * from right to left to the player.
+ **/
```
nit: only one `*` for closing :)

gkatsev commented on docs/guides/middleware.md:
```diff
@@ -5,14 +5,15 @@ Middleware is a Video.js feature that allows interaction with and modification o
+Middleware are functions that return an object with methods matching those on the `Tech`. There are currently a limited set of allowed methods that will be understood by middleware. These are: `buffered`, `currentTime`, `setCurrentTime`, `duration`, `seekable`, `played`, `play` and `pause`.
 
-These allowed methods are split into two categories: `getters` and `setters`. Setters will be called on the `Player` first and run through middleware(from left to right) before calling the method, with its arguments, on the `Tech`. Getters are called on the `Tech` first and are run though middleware(from right to left) before returning the result to the `Player`.
+These allowed methods are split into three categories: `getters`, `setters` and `mediators`. Setters will be called on the `Player` first and run through middleware(from left to right) before calling the method, with its arguments, on the `Tech`. Getters are called on the `Tech` first and are run though middleware(from right to left) before returning the result to the `Player`.
```
Probably. Would be nice to get a one-sentence description, though, if we can.

gkatsev commented on src/js/video.js:
```diff
@@ -333,6 +333,31 @@ videojs.registerTech = Tech.registerTech;
+    enumerable: true
+  });
+} else {
+  videojs.middleware.TERMINATOR = TERMINATOR;
```
this will fail because `videojs.middleware` isn't an object.
Should be `videojs.middleware = {TERMINATOR}`

gkatsev commented:
Oh, the commit message should probably be more like `feat: add mediator middleware type for play()`

ldayananda commented on docs/guides/middleware.md:
```diff
@@ -5,14 +5,16 @@ Middleware is a Video.js feature that allows interaction with and modification o
+Middleware are functions that return an object with methods matching those on the `Tech`. There are currently a limited set of allowed methods that will be understood by middleware. These are: `buffered`, `currentTime`, `setCurrentTime`, `duration`, `seekable`, `played`, `play`, `pause` and `paused`.
 
-These allowed methods are split into two categories: `getters` and `setters`. Setters will be called on the `Player` first and run through middleware(from left to right) before calling the method, with its arguments, on the `Tech`. Getters are called on the `Tech` first and are run though middleware(from right to left) before returning the result to the `Player`.
+These allowed methods are split into three categories: `getters`, `setters` and `mediators`. Setters will be called on the `Player` first and run through middleware(from left to right) before calling the method, with its arguments, on the `Tech`. Getters are called on the `Tech` first and are run though middleware(from right to left) before returning the result to the `Player`. Mediators are call ...
```
@gkatsev Added a sentence for Mediators, but it's more confusing in my opinion than reading the full section

gkatsev replied:
🤔  I think this is ok. Maybe @videojs/core-committers has any other thoughts?

misteroneill replied:
I think this paragraph should be split up. I propose breaking it into three distinct sections. Also, made some grammatical improvements.

----------

These allowed methods are split into three categories: getters, setters, and mediators.

### Middleware Setters
Setters will be called on the `Player` first and run through middleware (from left to right) before calling the method, with its arguments, on the `Tech`.

### Middleware Getters
Getters are called on the `Tech` first and are run though middleware (from right to left) before returning the result to the `Player`.

### Middleware Mediators
Mediators are called on the `Player` first, run through middleware (from left to right), then called on the `Tech`. The result is returned to the `Player` unchanged, while calling the middleware from right to left. For more information on mediators, check out the [mediator section](#termination-and-mediators).

misteroneill commented on src/js/utils/promise.js:
```diff
@@ -9,7 +9,7 @@
  */
 export function isPromise(value) {
-  return value !== undefined && typeof value.then === 'function';
+  return value !== undefined && value !== null && typeof value.then === 'function';
```
Mabye we should just do `return Boolean(value) && typeof value.then === 'function';`

ldayananda replied:
it's nice to be explicit here too 🤔

misteroneill commented on src/js/video.js:
```diff
@@ -325,8 +325,39 @@ videojs.getTech = Tech.getTech;
+if (!browser.IS_IE8 && Object.defineProperty) {
+  Object.defineProperty(videojs, 'middleware', {
+    value: {},
+    writeable: false,
```
I believe `writable: false` is the default, so this shouldn't be necessary.

ldayananda replied:
We talked about this elsewhere, but since having `writeable: false` is very explicit, I plan to keep it 😄
